active/0051-strategies.md

```diff
@@ -0,0 +1,247 @@
++ **Assembler** - Responsible for taking the resolved tree and applying – default or user provided – strategies to the tree.
+
+# Detailed design
+
+The detailed design is separated in various sections so that it is easier for a reader to understand.
+
+## Assembler
+
+`Assembler` is similar to a car assembly line - starting with a metal frame, making changes to it along the way, adding parts and at the end, there's an assembled car.
+
+It does not make any assumptions about how the final output should be constructed. Instead, it relies on strategies to tell it what the final output should look like. This is rather important as it enforces clear separation of concerns.
+
+Assembler must have the following interface:
+
+```typescript
+interface Assembler {
+  constructor(inputTree: BroccoliTree, strategies: Strategy[]);
+  toTree(): BroccoliTree;
+}
+```
+
+`toTree` method would gather all the trees from the passed in `strategies`, and merge them into one final broccoli tree.
+
+## Strategies
+
+Strategy is an extensibility primitive. It gives you granular control over the final output. It could be used in many different ways (we are going to go over use cases below).
+
+Strategy must have the following interface:
+
+```typescript
+interface Strategy {
+  constructor(options: any);
+  toTree(assember: Assembler, inputTree: BroccoliTree): BroccoliTree;
+}
+```
+
+`inputTree` will have the following structure:
+
+```sh
+bundler:js:input/
+├── addon-modules
+├── the-app-name-folder
+├── node_modules
+└── vendor
+```
+
+where `addon-modules` contains dependencies from Ember add-ons; `the-app-name-folder` contains Ember application code; `node_modules` contains node dependencies and `vendor` contains other dependencies.
+
+Note, that  `toTree` method must return a broccoli tree.
+
+### Concatenation Strategy
+
+One of the strategies that we already use now to produce `assets/app-name.js` and `assets/vendor.js`. It takes a broccoli tree and returns a new concatenated broccoli tree.
+
+```javascript
+const concat = require('broccoli-concat');
+
+class ConcatenationStrategy {
+  constructor(options) {
+    this.options = options;
+  }
+
+  toTree(assembler, inputTree) {
+    return concat(inputTree, this.options);
+  }
+}
+```
+
+Where `this.options` mimics [`broccoli-concat` options](https://github.com/broccolijs/broccoli-concat#advanced-usage).
+
+### Example of possible strategies
+
+Below are some of the strategies that could be implemented.
+
+#### Assets Split Strategy
+
+Other strategy that proved very useful is an ability to separate assets. One of the applications would be to separate "static" assets from application code. They are called "static" because they don't change very often and could stay cached by CDN throughout several deployments (`ember.js`, `jQuery` or third party dependencies). That way performance is less effected by deployments.
```

Comment:
First sentence needs to be restructured. I think this generally can be clearer.

Reply:
This example is not just an example, it is how bundling works today. Are you suggesting that it *could* be ported to a strategy or that it *be* ported?

Other items in the "examples" section here seem to be only for illustration, but this one I would expect will be written as a built-in.

Reply:
I suspect I need to rewrite this section to be more clear as you suggested. Current bundling works differently (unless I misunderstood what you meant) that is _all_ third party libraries end up in one `vendor.js` file as opposed to being bundled separately.

I think this strategy could be provided out of the box by Ember CLI.